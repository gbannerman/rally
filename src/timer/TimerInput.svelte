<script lang="ts">
  import { insertSeperator, removeSeperator } from "./timerService";

  export let value = "";

  let timerInput: HTMLInputElement = null;

  const widthMap: { [key: number]: string } = {
    0: "width: 10px; margin-right: 2px; padding-left: 0px; right: 0px;",
    1: "width: 20px; margin-left: 21px;",
    2: "width: 37px; margin-left: 5px;",
    4: "width: 62px; margin-left: -22px;",
    5: "width: 79px; margin-left: -40px;",
  };

  $: inputWidth = widthMap[value.length];

  const handleInput = (event: Event) => {
    const newValue = (event.target as HTMLInputElement).value;
    const withoutSeperator = removeSeperator(newValue);
    value = insertSeperator(withoutSeperator);
  };

  const onSubmit = (event: Event) => {
    event.preventDefault();
    (document.activeElement as HTMLInputElement).blur();
  };

  const focusInput = () => {
    if (!timerInput) return;

    timerInput.focus();
    const endOfInput = value.length;
    timerInput.setSelectionRange(endOfInput, endOfInput);
  };
</script>

<div aria-label="Timer" class="TimerInput" on:click={focusInput}>
  <div class="timer placeholder">00:00</div>
  <form on:submit={onSubmit}>
    <input
      bind:this={timerInput}
      class="timer input"
      type="text"
      inputmode="numeric"
      maxlength="5"
      style={inputWidth}
      {value}
      on:input={handleInput}
      on:click={focusInput}
    />
  </form>
</div>

<style>
  .TimerInput {
    width: 100px;
    height: 40px;
    text-align: center;
  }

  input.timer {
    width: 42px;
    padding: 2px;
    margin: 5px;
    height: 30px;
    margin-left: 50px;
  }

  .timer {
    position: absolute;
    width: 100%;
    font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
    background-color: #e9e9e9;
    height: 40px;
    line-height: 20px;
    padding: 10px 10px;
    color: #606c6e;
    font-size: 30px;
    outline: 0 solid transparent;
    border: 0 solid transparent;
    border-radius: 4px;
    box-sizing: border-box;
  }
</style>
