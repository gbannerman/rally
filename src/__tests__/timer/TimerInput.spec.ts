import {render, fireEvent} from '@testing-library/svelte';
import userEvent from '@testing-library/user-event';

import TimerInput from '../../timer/TimerInput.svelte';

describe("TimerInput component", () => {
  it("renders correctly", () => {
    const { container } = render(TimerInput);
    
    expect(container).toMatchSnapshot();
  });

  it("clicking on timer focuses input", async () => {
    const { getByLabelText, getByRole } = render(TimerInput);

    await fireEvent.click(getByLabelText("Timer"));

    const input = getByRole("textbox");
    
    expect(input).toHaveFocus();
  });

  it("submitting input removes focus", async () => {
    const value = "";

    const { getByRole } = render(TimerInput);

    const input = getByRole("textbox");

    await userEvent.type(input, '129', { delay: 500 });
    await fireEvent.submit(input);
    
    expect(input).not.toHaveFocus();
  });

  it("formats value with seperator", async () => {
    const { getByRole } = render(TimerInput);

    const input = getByRole("textbox");

    await userEvent.type(input, '530', { delay: 500 });

    expect(input).toHaveValue("5:30");
  });

  it("only accepts 4 digits", async () => {
    const { getByRole } = render(TimerInput);

    const input = getByRole("textbox");

    await userEvent.type(input, '20000', { delay: 10 });

    expect(input).toHaveValue("20:00");
  });

  it("input does not hide placeholder when no value", () => {
    const { getByRole } = render(TimerInput);

    const input = getByRole("textbox");

    expect(input).toHaveStyle("width: 10px; margin-right: 2px; padding-left: 0px; right: 0px;");
  });

  it("input hides first digit of placeholder when value length is 1", async () => {
    const { getByRole } = render(TimerInput);

    const input = getByRole("textbox");

    await userEvent.type(input, '2', { delay: 10 });

    expect(input).toHaveStyle("width: 20px; margin-left: 21px;");
  });

  it("input hides first digit of placeholder when digit entered", async () => {
    const { getByRole } = render(TimerInput);

    const input = getByRole("textbox");

    await userEvent.type(input, '2', { delay: 10 });

    expect(input).toHaveStyle("width: 20px; margin-left: 21px;");
  });

  it("input hides two digits of placeholder when two digits entered", async () => {
    const { getByRole } = render(TimerInput);

    const input = getByRole("textbox");

    await userEvent.type(input, '21', { delay: 10 });

    expect(input).toHaveStyle("width: 37px; margin-left: 5px;");
  });

  it("input hides three digits of placeholder when three digits entered", async () => {
    const { getByRole } = render(TimerInput);

    const input = getByRole("textbox");

    await userEvent.type(input, '214', { delay: 10 });

    expect(input).toHaveStyle("width: 62px; margin-left: -22px;");
  });

  it("input hides four digits of placeholder when four digits entered", async () => {
    const { getByRole } = render(TimerInput);

    const input = getByRole("textbox");

    await userEvent.type(input, '2145', { delay: 10 });

    expect(input).toHaveStyle("width: 79px; margin-left: -40px;");
  });
});
