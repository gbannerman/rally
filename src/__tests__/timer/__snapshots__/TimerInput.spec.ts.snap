// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimerInput component renders correctly 1`] = `
<body>
  <div>
    <div
      aria-label="Timer"
      class="TimerInput svelte-jd5xj"
    >
      <div
        class="timer placeholder svelte-jd5xj"
      >
        00:00
      </div>
       
      <form>
        <input
          class="timer input svelte-jd5xj"
          inputmode="numeric"
          maxlength="5"
          style="width: 10px; margin-right: 2px; padding-left: 0px; right: 0px;"
          type="text"
        />
      </form>
    </div>
  </div>
</body>
`;
